Add gg.cmd run-pr xyz (#115)
JabRef/blog.jabref.org: _posts/2025-05-31-run-pr.md

---
title: How to try out any JabRef pull request
---

JabRef gets many contributions by external contributors in the form of [pull requests](https://github.com/jabref/jabref/pulls).

> Curious how the feature or fix behaves?
> You now can try out and give feedback easily.

Reviewer: ```suggestion
Want to try out the feature or fix?
You can now do so in a few simple steps, and help the team by giving early feedback!
```
or
```suggestion
Want to try out the feature or fix?
You can now do so in a few simple steps, give early feedback to the team!
```
or (since it's an external contributor)
```
or
```suggestion
Want to try out the feature or fix?
You can now do so in a few simple steps, give early feedback!
```

For instance, to try out the "[initial implementation of walkthrough](https://github.com/JabRef/jabref/pull/13182)", one needs to execute following command:

- Windows: `.\gg.cmd just run-pr 13182`
- macOS/Linux: `sh ./gg.cmd just run-pr 13182`

## Preparation

To make this work, one needs to have [`gg.cmd`](https://github.com/eirikb/gg#ggcmd) available in JabRef's source directory.

This can be achieved by using `gg.cmd`, `JBang`, and scripts offered by JabRef:

1. [Download `gg.cmd`](https://github.com/eirikb/gg/releases/latest/download/gg.cmd)
2. Instruct `JBang` to trust JabRef: Execute `.\gg.cmd jbang trust add https://github.com/JabRef/jabref/`.\
   ![gg initialization](../img/gg-init-1.png)\
   ![gg initialization](../img/gg-init-2.png)
3. Clone JabRef into a folder.

>    As default, a system-wide available temporary folder is used.

Reviewer: ```suggestion
   By default, a system-wide available temporary folder is used.
```

>    (In "Advanced Hints" below, we will show how to use a specific folder.)

Reviewer: ```suggestion
   (In "Advanced Hints" below, we mention steps on how to use a specific folder.)
```

   Execute `.\gg.cmd jbang https://github.com/JabRef/jabref/blob/main/.jbang/CloneJabRef.java`.\
   ![gg-jabref-clone](../img/gg-jabref-clone.png)
4. After it finishes, it will output the directory:

   ```text
   JabRef code available at: C:\Users\WDAGUtilityAccount\AppData\Local\Temp\jabref
   ```

5. Go to the directory: `cd C:\Users\WDAGUtilityAccount\AppData\Local\Temp\jabref` (Note - replace "`WDAGUtilityAccount`" with your username).
6. Move `gg.cmd` to here: `move C:\Users\WDAGUtilityAccount\Downloads\gg.cmd .`. (Note - replace "`WDAGUtilityAccount`" with your username. The path also needs to be changed if you have downloaded `gg.cmd` anywhere else.)

> 7. Now you can run a PR using `.\gg.cmd just run-pr 13182` (replace `13182` with the PR number of whichever PR you want to try out). Initially, this might take several minutes, as the build environment has to be initialized and JabRef is build completely from scratch. In later runs the run will be faster.

Reviewer: ```suggestion
7. Now you can run a PR using `.\gg.cmd just run-pr 13182` (replace `13182` with the PR number of whichever PR you want to try out). Initially, this might take several minutes, as the build environment has to be initialized and JabRef has to be built completely from scratch. In subsequent runs, it is much faster.
```

## Advanced Hints

### Using another directory

In step 3, a temporary folder was used.
You can also use a different directory.
For example, following command clones into a directory `jabref` located in the current folder:

```cmd
.\gg.cmd jbang https://github.com/JabRef/jabref/blob/main/.jbang/CloneJabRef.java jabref
```

In general, any (non-existing) directory can be used.
To clone into `c:\git-repositories\jabref`, use following command:

```cmd
.\gg.cmd jbang https://github.com/JabRef/jabref/blob/main/.jbang/CloneJabRef.java c:\git-repositories\jabref
```

### Linux and macOS

For Linux and macOS, use `sh ./gg.cmd` instead of `.\gg.cmd`.

### Running `JabKit`

You can run other parts of JabRef. For instance for `JabKit`:

```cmd
.\gg.cmd just run-jabkit --help
```

> ### Using a Windows Sandbox

Reviewer: Oh nvm it's already under advanced

---
title: Try out any JabRef pull request
---

> Thanks to the strong ecosystem of Java, one can easily try out (nearly) any JabRef pull request.

Reviewer 2: Some more context, for example 
you want to try out a feature in development or fix and give feedback

Author: Done.

1. [Download `gg.cmd`](https://github.com/eirikb/gg/releases/latest/download/gg.cmd)
2. Execute `.\gg.cmd jbang https://github.com/JabRef/jabref/blob/main/.jbang/CloneJabRef.java`. This will clone JabRef into a temporary folder.

    1. Initialization\
       ![gg initialization](../img/gg-init.png)
    2. JBang will ask for trust - please trust the script with <kbd>2</kbd> and then <kbd>Enter</kbd>\
       ![JBang confirmation](../img/gg-jbang-confirmation.png)
    3. Clone happens\
       ![gg-jabref-clone](../img/gg-jabref-clone.png)

3. After it finishes, it will output the directory:

   ```text
   JabRef code available at: C:\Users\koppor\AppData\Local\Temp\jabref
   ```

4. Change into the directory: `cd C:\Users\koppor\AppData\Local\Temp\jabref`. Replace `koppor` by your username.
5. Move `gg.cmd` to here: `move C:\Users\koppor\Downloads\gg.cmd .`. Replace `koppor` by your username.
6. Now you can run a PR using `.\gg.cmd just run-pr 13182`. Initially, this might take several minutes, as the build environment has to be initialized and JabRef is build completely from scratch. You will also need to tell JBang to trust JabRef's code. In later runs the run will be faster.

> ## Trouble Shooting

Reviewer: ```suggestion
## Troubleshooting
```

For Linux and macOS, use `sh ./gg.cmd` instead of `.\gg.cmd`.

You might get "error: No justfile found"

```monospace
just gui
error: No justfile found
error: Recipe `run-pr` failed on line 21 with exit code 1
Unable to execute
```

Then, the pull request has not merged recent `main`.

> You need to "manually" start:

Reviewer: Does this take care of taking a pull automatically in that case?

Reviewer: So we need to add another instruction to update their local main before the next command? As if this error occurred, most probably their main is also outdated in state

Author: The pull request is out of date. Example:

https://github.com/JabRef/jabref/pull/13182

Comment:

```markdown
Dear @Yubo-Cao, You did not merge that latest `upstream/main` yet. Please do that as it ensures you have the latest developments in. Thank you.
```

---

Reading this discussion, I just delete this paragraph and we ensure that all PRs have recent main merged.

Reviewer: I think, we may be talking about different things.
Your context - the pull request checkout does not have updates from upstream/main
My context - since the PR is merged, normal instructions give error, so here are instructions are how you checkout the feature. But for that, the local main has to be updated as per upstream/main.

Author: Implemented in https://github.com/JabRef/jabref/pull/13195

4. After it finishes, it will output the directory:

   ```text

>    JabRef code available at: C:\Users\WDAGUtilityAccount\AppData\Local\Temp\jabref

Reviewer: Is it okay to have a new folder under users
We know it's perfectly fine
But from the perspective of a third person, that's like messing with the users structure of the system

Author: The folder is the temporary folder for each operating system...

We could change for Windows to let users specify the folder. These are more commands, therefore I did not do.

Ah, maybe under advanced as some paragraph?

Reviewer: if it is auto-generated, it is fine. I thought you changed it from self (koppor)

Author: It's just the user of "Windows Sandbox" :)
